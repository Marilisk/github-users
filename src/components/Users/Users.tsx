import { FC, useEffect } from "react"
import { IUser } from "../../types/usersTypes"
import { UserItem } from "./UserItem/UserItem"
import { useAppDispatch, useAppSelector } from "../../redux/store/hooks"
import { Preloader } from "../../assets/Preloader/Preloader"
import { createPortal } from "react-dom"
import { UserPopup } from "./UserPopup/UserPopup"
import { callPopup } from "../../redux/store/usersSlice"

interface IUsersProps {
    items: IUser[]
}

export const Users: FC<IUsersProps> = ({ items }) => {

    const isLoading = useAppSelector(s => s.users.users.loadingStatus === 'loading')
    const query = useAppSelector(s => s.users.query)
    const dispatch = useAppDispatch()
    const popup = useAppSelector(s => s.users.popup)
    const showPopup = popup?.opened

    const handleClick = (e: MouseEvent) => {
        const element = (e.target as HTMLElement).closest('[id]') as HTMLElement
        let id
        if (element) { id = element.id }
        dispatch(callPopup(id))
    }

    useEffect(() => {
        window.addEventListener('click', handleClick)
        return () => window.removeEventListener('click', handleClick)
    })

    if (!query?.length) return null
    if (isLoading) return <Preloader />
    if (!items.length && query.length && !isLoading) return <h1>нет результатов...</h1>

    return <div>
        {items.map((user => <UserItem key={user.id} user={user} />))}
        {showPopup && createPortal(
            <UserPopup userId={popup.id} />,
            document.body)}
    </div>
}