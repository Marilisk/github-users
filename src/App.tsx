import { Users } from './components/Users/Users'
import { useAppSelector } from './redux/store/hooks'
import { SearchBar } from './components/SearchBar/SearchBar'
import { Pagination } from './components/Pagination/Pagination'

function App() {

  const users = useAppSelector(s => s.users.users.items)
  const total = useAppSelector(s => s.users.users.totalUsersCount)


  return (
    <>
      <SearchBar />
      {!!total && !!users.length &&
        <>
          <h1>Результатов: {total}</h1>
          <Pagination />
        </>
      }
      <Users items={users} />
    </>
  )
}

export default App
