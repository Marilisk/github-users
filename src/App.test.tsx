import App from './App.tsx';
import { render, screen } from './utils/test-utils'
import * as reduxHooks from 'react-redux'
import { Users } from './components/Users/Users.tsx';
import { mockUser } from './components/Users/UserItem/UserItem.test.tsx';

vitest.mock('react-redux')
const mockUsers = [mockUser]


it('the input is visible', () => {
    vitest.spyOn(reduxHooks, 'useSelector').mockReturnValue([])
    render(<App />)
    const placeholder = screen.queryByText(/введите логин/i)
    expect(placeholder).toBeDefined()
})

it('no results: not to be in the component by default', () => {
    vitest.spyOn(reduxHooks, 'useSelector').mockReturnValue([])
    render(<App />)
    const placeholder = screen.queryByText(/нет результатов.../i)
    expect(placeholder).not.toBeInTheDocument()
})

it('should create App with empty users', () => {
    const component = render(<App />)
    expect(component).toMatchSnapshot()
})

it('should show users', () => {
    const component = render(<Users items={mockUsers} />)
    expect(component).toBeDefined()
})



