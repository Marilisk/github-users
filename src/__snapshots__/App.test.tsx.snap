// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create App with empty users 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_wrap_3bb1d7"
      >
        <div
          class="_bar_3bb1d7"
        >
          <input
            data-testid="searchInput"
            placeholder="введите логин"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="_hidden_3e9ab6"
      >
        <button
          class="_pass_3e9ab6"
          type="button"
        >
          <span>
            сортировать по возрастанию
          </span>
        </button>
        <button
          class="_act_3e9ab6"
          type="button"
        >
          <span>
            сортировать по убыванию
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_wrap_3bb1d7"
    >
      <div
        class="_bar_3bb1d7"
      >
        <input
          data-testid="searchInput"
          placeholder="введите логин"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="_hidden_3e9ab6"
    >
      <button
        class="_pass_3e9ab6"
        type="button"
      >
        <span>
          сортировать по возрастанию
        </span>
      </button>
      <button
        class="_act_3e9ab6"
        type="button"
      >
        <span>
          сортировать по убыванию
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should create users 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should show  users 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should show users 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
